Add a MOOSE-aware PR review agent skill (#33139)
idaholab/moose: .claude/skills/moose-pr-review/SKILL.md

---

# Reviewing a MOOSE Pull Request

MOOSE is safety-relevant software with a strict software-quality program (NQA-1). A review
is not just "does the code look right" - it is the gate that enforces the Code Standard, the
SQA traceability between tests and requirements, and the documentation that lets the next
person understand the change without asking the author. This skill applies the same bar a
Change Control Board reviewer would, expressed as concrete, checkable steps.

This skill is the MOOSE-specific layer. It does **not** re-implement generic bug hunting or
security analysis - those are covered well by the built-in `/code-review` and
`/security-review` skills. Run or recommend those for deep correctness/security; spend this
skill's attention on what is unique to MOOSE and what CI cannot auto-check.

## What CI already enforces (do not nitpick by hand)

Formatting is checked automatically on every PR, so manual review time should not be spent on
it. If you see violations, note them once and point to the fix command rather than listing
each instance:

- C++ formatting: `git clang-format <base>` (uses `.clang-format`)
- Python formatting: `black .` (uses `pyproject.toml`)

> - Trailing whitespace / tabs, and the "every commit references an issue" check

Reviewer: ```suggestion
- No trailing whitespace / tabs, and the "every commit references an issue" check
```

This review depends on CodeGraph to detect code that reinvents functionality MOOSE already has.
That is the single most valuable thing this review catches against MOOSE's "use existing
functionality, don't reimplement it" rule (AGENTS.md, Simplicity First), and it is something
grep cannot do reliably across a 100k-symbol codebase. CodeGraph is therefore a hard
prerequisite, not an optimization.

Before anything else, confirm the index exists: check for a `.codegraph/` directory at the
repository root (or run `codegraph status`).

If it is missing, **stop and surface it to the user - do not fall back to grep and proceed.**
Tell them CodeGraph is required for a MOOSE review and how to set it up:

```bash
codegraph install   # one-time: register the CodeGraph MCP server with this agent
codegraph init      # build the index for this repo (run from the repo root; takes a few minutes)
```

Only continue once the index is present. Throughout the review, prefer the CodeGraph MCP tools
(`codegraph_explore`, `codegraph_search`, `codegraph_node`, `codegraph_callers`) over grep/Read
whenever the question is "does this already exist?", "what calls this?", or "what does changing

> this affect?" See `references/codegraph.md` for the exact query patterns.

Reviewer: I wonder if we should put this same instruction set somewhere else just for general agent coding sessions, without the hard failure if codegraph doesn't exist?

Author: I added a moose codegraph skill. It encourages the use of codegraph but doesn't enforce it. The review skill references it and enforces its usage.

## Step 1 - Identify what you are reviewing

Determine the target and compute the diff. Two cases:

**A GitHub PR** (user gives a number or URL, or asks to review a specific PR):

```bash
gh pr view <num> --json title,body,baseRefName,headRefName,author,files,commits,state,statusCheckRollup
gh pr diff <num>
```

`baseRefName` is the merge target (MOOSE PRs target `next`). Use the PR body to read the
author's stated Reason / Design / Impact (the PR template fields).

**A local branch** (the default when no PR is named):

```bash
# Review the branch against the point it was cut from. <base> is the integration branch the

> # work branched from (its upstream tracking branch, e.g. origin/devel). Use three dots so the

Reviewer: I dont think upstream tracking branch is correct in terms of that corresponding to `base`. E.g.
```
(base) lindad@INL429698:~/projects/moose-petsc-dbg(fix-iteration-ghosting)$ git rev-parse --abbrev-ref --symbolic-full-name @{u}
origin/fix-iteration-ghosting
```

- Python (`*.py`) - Code Standard (Python section)
- Test specs (files literally named `tests`) and inputs (`*.i`) - SQA testing checks
- Markdown under `doc/content/` - documentation checks
- Build/config (`Makefile`, `*.mk`, `configure.ac`) - flag for extra scrutiny

## Step 2 - Gather MOOSE context

- **Linked issue(s).** Read the issue(s) the commits reference - they are your spec for judging
  whether the change is complete and correctly scoped. (CI already enforces that a reference
  exists, so don't spend review time policing its presence.)
- **Newly registered objects.** Find new user-facing objects in the diff:
  `git diff <base>...HEAD | grep -E '^\+.*register(MooseObject|.*Action)\('`.
  Each new object drives both a documentation check (Step 3E) and a testing check (Step 3D).
- **Scope.** Is the change surgical and focused, or does it bundle unrelated edits ("rider"
  commits)? MOOSE reviewers may legitimately ask to split a sprawling PR. Note mixed concerns.

## Step 3 - Review across the MOOSE dimensions

Work top-down: assess the high-level design first (a sound design with rough edges is fixable;

> a wrong design is not), then the details. For each dimension, the highest-value checks are

Reviewer: ```suggestion
a wrong design may not be), then the details. For each dimension, the highest-value checks are
```

summarized here; the reference files hold the full checklists - read the relevant one when the
diff touches that area.

**A. Design, scope, and user interface** (`reviewing.md`)
- Is the high-level design sound, and proportionate scrutiny applied (changes affecting many
  users or that others will build on deserve more)?
- Scrutinize the **user interface**: new input parameters, their names, defaults, and whether
  `validParams` documentation strings make them self-explanatory. Interface mistakes are
  expensive because users depend on them.
- Will the code be understandable outside the PR context, from comments and docs alone?

**B. Reuse and duplication (CodeGraph-driven - vital)** -> read `references/codegraph.md`
- For every new function, class, or method the PR introduces, ask CodeGraph whether MOOSE
  already provides it. Reinventing an existing utility, algorithm, or object is one of the most

>   common and most important things to catch - it bloats the codebase and diverges behavior.

Reviewer: ```suggestion
  common and most important things to catch - it bloats the codebase, diverges behavior, and
  increases maintenance costs.
```

  `codegraph_explore` the concept to surface existing equivalents. If a near-duplicate exists,
  flag it as required and point the author to reuse or extend it (a short extension/bugfix of
  existing code is preferred over a long reimplementation).
- Also use `codegraph_callers` / `impact` on changed existing symbols to judge blast radius
  (this feeds the scope/scrutiny assessment in A) and to find call sites the PR should update.

**C. MOOSE Code Standard** -> read `references/code-standard.md`
- Top semantic checks CI misses: strict **const-correctness** (logically-const-but-unmarked is
  a design issue), **access control** (default `private`; minimal `public`; ordered
  public->protected->private), C++-over-C constructs (`nullptr`, `static_cast`, `enum class`,
  RAII, `make_unique`/`make_shared`), header hygiene (forward-declare in headers, include in
  `.C`), `override`, virtual destructors, and the naming conventions.

**D. Tests and SQA** -> read `references/testing-sqa.md`
- Every new feature/behavior needs a test, and **every test is a requirement**. Each test
  block must carry an unambiguous `requirement`, a `design` (markdown path), and `issues`

>   (`#numbers`). Error paths deserve `RunException` tests. Diff-based tests need gold files.

Reviewer: I think we need some more nuance here. I am writing more and more `unit` tests these days for which we we don't (currently) have the `requirement`,`design`, `issues` requirements. @cticenhour is this something that we're going to need to think about? SQA for `unit` tests? I have PRs that only write unit tests

Reviewer: resolving for now. I've opened https://github.com/idaholab/moose/issues/33325

- Missing or under-specified tests are the most common required findings - check this carefully.

**E. Documentation** -> read `references/documentation.md`
- Each new MooseObject needs `addClassDescription(...)` in `validParams()` **and** a markdown
  stub page mirroring the source path (e.g. `src/kernels/Foo.C` ->

>   `framework/doc/content/source/kernels/Foo.md`). A missing stub page fails the docs build.

Reviewer: ```suggestion
  `doc/content/source/kernels/Foo.md`). A missing stub page fails the docs build.
```

- Significant changes warrant a newsletter entry (current month, e.g.
  `modules/doc/content/newsletter/2026/2026_06.md`).

**F. Correctness and security** (defer, don't duplicate)
- For logic bugs, edge cases, and efficiency, run the built-in `/code-review` (or recommend it).
- For anything security-sensitive (parsing untrusted input, file/system access, memory safety),
  run or recommend `/security-review`. Memory-management-heavy code may warrant a valgrind

>   testing recipe; loose tolerances often break parallel testing - mention these when relevant.

Reviewer: ```suggestion
  testing recipe; loose solve tolerances often break parallel testing - mention these when relevant.
```
